cmd/soroban-rpc/internal/db/migration.go

```diff
@@ -0,0 +1,174 @@
+	var result *MigrationLedgerRange
+	for _, m := range mm {
+		r := m.ApplicableRange()
+		if r != nil {
+			if result == nil {
+				result = r
+			} else {
+				result.firstLedgerSeq = min(r.firstLedgerSeq, result.firstLedgerSeq)
+				result.lastLedgerSeq = max(r.lastLedgerSeq, result.lastLedgerSeq)
+			}
+		}
+	}
+	return result
+}
+
+func (mm multiMigration) Apply(ctx context.Context, meta xdr.LedgerCloseMeta) error {
+	var err error
+	for _, m := range mm {
+		if localErr := m.Apply(ctx, meta); localErr != nil {
+			err = errors.Join(err, localErr)
+		}
+	}
+	return err
+}
+
+func (mm multiMigration) Commit(ctx context.Context) error {
+	var err error
+	for _, m := range mm {
+		if localErr := m.Commit(ctx); localErr != nil {
+			err = errors.Join(err, localErr)
+		}
+	}
+	return err
+}
+
+func (mm multiMigration) Rollback(ctx context.Context) error {
+	var err error
+	for _, m := range mm {
+		if localErr := m.Rollback(ctx); localErr != nil {
+			err = errors.Join(err, localErr)
+		}
+	}
+	return err
+}
+
+// guardedMigration is a db data migration whose application is guarded by a boolean in the meta table
+// (after the migration is applied the boolean is set to true, so that the migration is not applied again)
+type guardedMigration struct {
+	guardMetaKey    string
+	db              *DB
+	migration       MigrationApplier
+	alreadyMigrated bool
+}
+
+func newGuardedDataMigration(ctx context.Context, uniqueMigrationName string, factory migrationApplierFactory, db *DB) (Migration, error) {
+	migrationDB := db.Clone()
+	if err := migrationDB.Begin(ctx); err != nil {
+		return nil, err
+	}
+	metaKey := "Migration" + uniqueMigrationName + "Done"
+	previouslyMigrated, err := getMetaBool(ctx, migrationDB.SessionInterface, metaKey)
+	if err != nil && !errors.Is(err, ErrEmptyDB) {
+		migrationDB.Rollback()
+		return nil, err
+	}
+	latestLedger, err := NewLedgerEntryReader(db).GetLatestLedgerSequence(ctx)
+	if err != nil && err != ErrEmptyDB {
+		migrationDB.Rollback()
+		return nil, fmt.Errorf("failed to get latest ledger sequence: %w", err)
+	}
+	applier, err := factory.New(migrationDB, latestLedger)
+	if err != nil {
+		migrationDB.Rollback()
+		return nil, err
+	}
+	guardedMigration := &guardedMigration{
+		guardMetaKey:    metaKey,
+		db:              migrationDB,
+		migration:       applier,
+		alreadyMigrated: previouslyMigrated,
+	}
+	return guardedMigration, nil
+}
+
+func (g *guardedMigration) Apply(ctx context.Context, meta xdr.LedgerCloseMeta) error {
+	if g.alreadyMigrated {
+		return nil
+	}
+	return g.migration.Apply(ctx, meta)
+}
+
+func (g *guardedMigration) ApplicableRange() *MigrationLedgerRange {
+	if g.alreadyMigrated {
+		return nil
+	}
+	return g.migration.ApplicableRange()
+}
+
+func (g *guardedMigration) Commit(ctx context.Context) error {
+	if g.alreadyMigrated {
+		return nil
+	}
+	err := setMetaBool(ctx, g.db.SessionInterface, g.guardMetaKey)
+	if err != nil {
+		return errors.Join(err, g.Rollback(ctx))
```

Comment:
I noticed we don't rollback when encountering `Apply()` errors. But the lack of rollbacks don't cause any issues because we exit the process as soon as encountering an error in `mustInitializeStorage()`. So perhaps we don't need to rollback here either? Or maybe the `Rollback()` function can be removed entirely since any error in the migration is treated as a fatal error.

Reply:
Good catch. We don't currently roll back, but we might in the future (depending on how we end up applying the migrations). We should clean-up the code (and maybe not apply the migrations in the daemon).